open-machine-learning-jupyter-book/assignments/llm/basic/transformer-architecture.ipynb

```diff
@@ -0,0 +1,726 @@
+    "def train_model(epochs, print_every=100):\n",
+    "    model.train()\n",
+    "    start = time.time()\n",
+    "    temp = start\n",
+    "    total_loss = 0\n",
+    "\n",
+    "    for epoch in range(epochs):\n",
+    "        for i, batch in enumerate(train_iter):\n",
+    "            src = batch.English.transpose(0, 1)\n",
+    "            trg = batch.French.transpose(0, 1)\n",
+    "            trg_input = trg[:, :-1]\n",
+    "            targets = trg[:, 1:].contiguous().view(-1)\n",
+    "            src_mask, trg_mask = create_masks(src, trg_input)\n",
+    "\n",
+    "            preds = model(src, trg_input, src_mask, trg_mask)\n",
+    "            optim.zero_grad()\n",
+    "            loss = F.cross_entropy(preds.view(-1, preds.size(-1)), targets, ignore_index=target_pad)\n",
+    "            loss.backward()\n",
+    "            optim.step()\n",
+    "            total_loss += loss.data[0]\n",
+    "\n",
+    "            if (i + 1) % print_every == 0:\n",
+    "                loss_avg = total_loss / print_every\n",
+    "                print(\"time = %dm, epoch %d, iter = %d, loss = %.3f, %ds per %d iters\" % (\n",
+    "                    (time.time() - start) // 60, epoch + 1, i + 1, loss_avg, time.time() - temp, print_every))\n",
+    "                total_loss = 0\n",
+    "                temp = time.time()\n",
+    "\n",
+    "## Train the model\n",
+    "train_model(epochs=10)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Test the trained model:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def translate(model, src, max_len=80, custom_string=False):\n",
+    "    model.eval()\n",
+    "    if custom_string == True:\n",
+    "        src = tokenize_en(src)\n",
+    "    sentence = Variable(torch.LongTensor([[EN_TEXT.vocab.stoi[tok] for tok in sentence]])).cuda()\n",
+    "    src_mask = (src != input_pad).unsqueeze(-2)\n",
+    "    e_outputs = model.encoder(src, src_mask)\n",
+    "    outputs = torch.zeros(max_len).type_as(src.data)\n",
+    "    outputs[0] = torch.LongTensor([FR_TEXT.vocab.stoi['<sos>']])\n",
+    "\n",
+    "    for i in range(1, max_len):\n",
+    "        trg_mask = np.triu(np.ones((1, i, i), k=1).astype('uint8'))\n",
+    "        trg_mask = Variable(torch.from_numpy(trg_mask) == 0).cuda()\n",
+    "        out = model.out(model.decoder(outputs[:i].unsqueeze(0), e_outputs, src_mask, trg_mask))\n",
+    "\n",
+    "        out = F.softmax(out, dim=-1)\n",
+    "        val, ix = out[:, -1].data.topk(1)\n",
+    "        outputs[i] = ix[0][0]\n",
+    "\n",
+    "        if ix[0][0] == FR_TEXT.vocab.stoi['<eos>']:\n",
+    "            break\n",
+    "\n",
+    "    return ' '.join([FR_TEXT.vocab.itos[ix] for ix in outputs[:i]])\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Acknowledgments\n",
```

Comment:
I think the need to thank the GPT needs to be considered

Reply:
removed this code